### PR TITLE
Replacing header files and common blocks with Fortran modules.

### PR DESCRIPTION
**Pull request by matsbn**
This is a rather substantial restructuring of the code to replace header files and common blocks with Fortran modules. I have tested the code in BLOM standalone mode and coupled to NorESM in NOIIA and N1850frc2 compsets. In my tests, the modified code produces bit-reproducible results with current master in NorESM startup, hybrid and continue types of run and gives no execution error when built with DEBUG set to TRUE in env_build.xml.

### CODE REVIEW

**Comment by matsbn** [comment]
@JorgSchwinger found an iHAMOCC issue with this PR when PnetCDF was used. His proposed correction has been added to the PR in the commit https://github.com/NorESMhub/BLOM/pull/61/commits/7d0deb4fe8ae9f93cc5bad116146fec91b0f298c.

**Review by JorgSchwinger — APPROVED** [review]

File: ben02/inicon_ben02.F [hunk 1]
```diff
@@ -65,8 +71,21 @@ subroutine icaux_ben02
           ticem(i,j)=tsrfm(i,j)
           tml_tda(i,j)=temp(i,j,1)+t0deg
           sml_tda(i,j)=saln(i,j,1)
-          alb_tda(i,j)=albs_f*ficem(i,j)+albw_d*(1.-ficem(i,j))
```
Comment (TomasTorsvik) [inline ben02/inicon_ben02.F:68]: change of variable: "albw_d" -> "albdif" (imported from mod_ben02)

File: ben02/mod_ben02.F [hunk 2]
```diff
@@ -98,6 +107,20 @@ module mod_ben02
 c
       integer ::
      .  ntda           ! accumulation number
+c
```
Comment (TomasTorsvik) [inline ben02/mod_ben02.F:110]: New functionality added in mod_ben02 (dfl, albw, alb, rnfins, rnfres, nrfets) ?

File: ben02/thermf_ben02.F [hunk 3]
```diff
@@ -371,18 +385,16 @@ subroutine thermf_ben02(m,n,mm,nn,k1m,k1n)
 c --- --- Accumulate the runoff in a reservoar to delay the discharge
 c --- --- into the ocean (by nrfets days approximately 1/e of runoff
 c --- --- added will by discharged).
-          rnfres(i,j)=rnfres(i,j)+rnf(i,j)
-          rnfflx(i,j)=rnfres(i,j)*rnf_fac
+          rnfres(i,j)=rnfres(i,j)+rnfins(i,j)
```
Comment (TomasTorsvik) [inline ben02/thermf_ben02.F:388]: Change runoff routine?

**Review by TomasTorsvik — APPROVED** [review]
I have scrolled through the diffs. Although it's quite long, it's mostly just replacements, some variables and subroutines changing names, and subroutines and common blocks converted to modules. There are some additions in ben02/mod_ben02.F that look more like change of functionality (new variables dfl, albw, alb, rnfins, rnfres). Could these potentially introduce answer-changing effects relative to noresm2.0?
Other than that I don't have further comments, so I think this can be merged with master.

File: ben02/mod_ben02.F (hunk 2 above)
Comment (TomasTorsvik) [reply ben02/mod_ben02.F:110]: Sorry, I wanted to point to the lines following line 110 ...

**Comment by AleksiNummelin** [comment]
I read through the changes and tried NHIST with f09_tn14 grid on Betzy, both BLOM and iHAMOCC were bit-identical. Can be resolved!

**Review by AleksiNummelin — APPROVED** [review]
I read through the changes and tried NHIST with f09_tn14 grid on Betzy, both BLOM and iHAMOCC were bit-identical. Can be resolved!

File: ben02/inicon_ben02.F (hunk 1 above)
Comment (matsbn) [reply ben02/inicon_ben02.F:68]: The variables you mention are declared in "phy/common_forc.h" of the current master (some are renamed) but are specific to the experiment configurations "ben02syn" and "ben02clim". Therefore I decided to declare them in modules contained in the "ben02" folder. They are not used when BLOM is coupled to NorESM. I have tested that the "ben02syn" and "ben02clim" configurations also produce bit-identical results.

File: ben02/thermf_ben02.F (hunk 3 above)
Comment (matsbn) [reply ben02/thermf_ben02.F:388]: It is just a change of variable name to make the forcing fields defined in ben02* experiments configurations and when coupled to NorESM more consistent. This does not change results or functionality.

File: ben02/mod_ben02.F (hunk 2 above)
Comment (matsbn) [reply ben02/mod_ben02.F:110]: My comment https://github.com/NorESMhub/BLOM/pull/61#discussion_r560225748 is related to this. No new functionality, just restructuring.

**Review by milicak — COMMENTED** [review]
Approved

**Pull request merged**